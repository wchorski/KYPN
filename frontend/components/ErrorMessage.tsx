import styled from 'styled-components';
import React from 'react';

import PropTypes from 'prop-types';
import { MdError } from 'react-icons/md';

const ErrorStyles = styled.div`
  padding: 2rem;
  background: white;
  margin: 2rem 0;
  border: 1px solid rgba(0, 0, 0, 0.05);
  border-left: 5px solid red;
  color: #931a33;

  p {
    color: black;
    margin: 0;
    font-weight: 100;
  }
  strong {
    margin-right: 1rem;
  }
`;

const ErrorMessage = ({ error }: any) => {

  if (!error || !error.message) return null;
  // console.log('error, ', error);

  if (error.networkError && error.networkError.result && error.networkError.result.errors?.length) {
    return error.networkError.result.errors.map((error: any, i: any) => (
      <ErrorStyles key={i}>
        <p data-testid="graphql-error">
          <strong> <MdError style={{ color: 'red' }} /> </strong>
          {error.message.replace('GraphQL error: ', '')}
        </p>
      </ErrorStyles>
    ));
  }
  return (
    <ErrorStyles>
      <p data-test="graphql-error">
        <strong> <MdError style={{ color: 'red' }} />  </strong>
        {error.code}
        {error.message.replace('GraphQL error: ', '')}
      </p>
    </ErrorStyles>
  );
};

ErrorMessage.defaultProps = {
  error: {},
};

ErrorMessage.propTypes = {
  error: PropTypes.object,
};

export default ErrorMessage;
