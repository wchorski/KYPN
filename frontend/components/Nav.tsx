import Link from 'next/link'
import React from 'react'
import styled from "styled-components";

const StyledNav = styled.nav`
  a{
    text-decoration: none;

    &:hover{
      text-decoration: underline;
    }
  }
`

export default function Nav() {
  return (
    <StyledNav>
      <Link href={`store`}> Store </Link>
      <Link href={`sell`}> sell </Link>
      <Link href={`orders`}> orders </Link>
      <Link href={`account`}> account </Link>
      <Link href={`blog`}> Blog </Link>
    </StyledNav>
  )
}
