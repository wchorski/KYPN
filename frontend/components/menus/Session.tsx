// import { useGlobalContext } from "../lib/useSessionContext";
import { gql, useQuery } from "@apollo/client";
import Link from "next/link";
import { MdAccountCircle } from "react-icons/md";
import styled from "styled-components";
import SignOutButton from "./SignOutButton";
import { LinkActive } from "./LinkActive";

// export const User = () => {
//   return (
//     <div>User</div>
//   )
// }

type Props = {
  session:any,
  label:string,
}

export function SessionBadge({ session, label, }: Props) {

  return (
    <StyledSessionBadge className={`toggle-menu button `} id="session-badge" aria-label='account menu link'>
      
      <Link href={`/account`} className="button account-toggle">
        <MdAccountCircle />
        <span>{label}</span>
      </Link>

      <ul className="sub-menu">
        <li className="name">{session?.name}</li>
        <li className="email">{session?.email}</li>
        <li><Link href={`/account`}> My Account </Link> </li>
        {session.isAdmin && (
            <li>
              <LinkActive href='/admin' className="button"> Admin Panel </LinkActive>
            </li>
          )}
        <li><SignOutButton /></li>
      </ul>
    </StyledSessionBadge>
  )
}


export function useUser() {
  // TODO GET THIS WORKING
  const { data } = useQuery(QUERY_USER_CURRENT)
  // console.log('++++++ useUser, ', data);
  return data?.authenticatedItem

  // const ctx = useGlobalContext()
  // return ctx?.session
}

export const QUERY_USER_CURRENT = gql`
  query AuthenticatedItem {
    authenticatedItem {
      ... on User {
        email
        id
        isAdmin
        name
        nameLast
        role {
          canManageCart
          canManageOrders
          canManageProducts
          canManageServices
          canManageRoles
          canManageUsers
          canSeeOtherUsers
          canManageTickets
          canManageEvents
          canManagePosts
          canManagePages
          canManageLocations
          canManageTags
          canManageCategories
          canManageSubscriptionPlans
          canManageSubscriptionItems
          canManageCoupons
        }
        tickets {
          event {
            id
            start
            summary
          }
          id
          status
        }
        subscriptions {
          id
          subscriptionPlan {
            id
          }
        }
        cart {
          id
          quantity
          product {
            id
            price
            name
            description
            image
            photo {
              image {
                publicUrlTransformed
              }
            }
          }
        }
      }
    }
  }
`

const StyledSessionBadge = styled.div`
  /* padding: 1em; */
  text-align: right;
  position: relative;
  margin-left: auto;
  background-color: var(--c-primary);
  border-radius: 100px;
  display: flex;
  align-items: center;
  color: var(--c-txt);
  border: none;

  svg{
    margin-right: .5rem;
    font-size: .8rem;
  }


  /* svg{
    font-size: 2rem;
  } */

  ul.sub-menu{
    opacity: 0;
    position: absolute;
    pointer-events: none;
    top: 100%;
    background-color: var(--c-body);
    display: flex;
    flex-direction: column;
    padding: .5em;
    box-shadow: var(--boxs-1);
    transform: translateY(10px);
    transition: all linear .1s;
    right: 0;
    z-index: 999;
    min-width: 20rem;
  }

  li{
    margin-bottom: 1em;
    padding: 0;
    text-align: right;
  }

  li.name{
    margin-bottom: 0;
    font-weight: bold;
  }
  li.email{
    font-size: small;
  }

  a{
    /* padding: 0 !important; */
    margin: 0;
  }

  &:hover, &:focus{
    ul{
      opacity: 1;
      pointer-events: all;
      transform: translateY(0px);
    }
  }
`