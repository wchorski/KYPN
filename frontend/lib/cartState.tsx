'use client'
import { createContext, useContext, useState } from "react";

const defaultCtx= {
  isOpen: false,
  setIsOpen: (isOpen:boolean) => {},
  toggleCart: () => {},
  closeCart: () => {},
  openCart: () => {},

}

const LocalStateContext = createContext(defaultCtx)

function CartStateProvider ({children}:{children: React.ReactNode}){

  const [isOpen, setIsOpen] = useState<boolean>(false)

  function toggleCart() {
    setIsOpen(!isOpen)
  }
  function closeCart() {
    setIsOpen(false)
  }
  function openCart() {
    setIsOpen(true)
  }

  return(
    <LocalStateContext.Provider 
      value={{isOpen, setIsOpen, toggleCart, closeCart, openCart}}
    >
      {children}
    </LocalStateContext.Provider>
  )
} 

function useCart(){
  const all = useContext(LocalStateContext)
  return all
}
export {CartStateProvider, useCart}