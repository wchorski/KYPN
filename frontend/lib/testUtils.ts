import casual from 'casual';
import { QUERY_PRODUCTS_COUNT } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  // __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  photo: {
    id: 'abc123',
    altText: 'dogs are best',
    image: {
      url: 'dog.jpg',
    },
  },
  name: 'dogs are best',
  description: 'dogs',
});

const fakeUser = (overrides: any | void) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  // @ts-ignore
  user: fakeUser(),
});

const fakeCartItem = (overrides: any | void) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  // @ts-ignore
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    // @ts-ignore
    this.store = {};
  }

  clear() {
    // @ts-ignore
    this.store = {};
  }

  getItem(key: any) {
    // @ts-ignore
    return this.store[key] || null;
  }

  setItem(key: any, value: any) {
    // @ts-ignore
    this.store[key] = value.toString();
  }

  removeItem(key: any) {
    // @ts-ignore
    delete this.store[key];
  }
}

function makePaginationMocksFor(length: any) {
  return [
    {
      request: { query: QUERY_PRODUCTS_COUNT },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
