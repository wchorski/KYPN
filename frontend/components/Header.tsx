import Link from 'next/link'
import React from 'react'
import {Nav} from './Nav'
import styled  from "styled-components";

const StyledLogo = styled.h1`
  background: var(--c-1);
  padding: .3em;
  margin: 0 .3em;
  transform: skew(-17deg);
  
  a{
    color: white;
    text-decoration: none;
  }
`

const StyledHeader = styled.header`
  .bar{
    border-bottom: solid 10px var(--c-1, grey);
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: center;
  }

  .sub-bar{
    border-bottom: solid 1px var(--c-2, darkgrey);
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: center;
  }
`

export default function Header() {
  return (
    <StyledHeader>

      <div className="bar">
        <StyledLogo>
          <Link href={`/`} > <h1> Sick Fits </h1> </Link>
        </StyledLogo>
        <Nav />
      </div>

      <div className="sub-bar">
        <input type="text" placeholder='search...'/>
      </div>


    </StyledHeader>
  )
}
