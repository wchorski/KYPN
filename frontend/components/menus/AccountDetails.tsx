import { useEffect, useRef, useState } from "react"
import { MdAutorenew, MdOutlineAccountBox, MdOutlineDownload, MdShop,  } from "react-icons/md"
import { HiOutlineTicket } from "react-icons/hi"
import styled from "styled-components"
import { Table } from "../elements/Table"
import { Booking, SubscriptionItem, User } from "../../lib/types"
import TicketsList from "../events/TicketsList"
import { datePrettyLocalDay, datePrettyLocalTime } from "../../lib/dateFormatter"
import Link from "next/link"
import { FiEdit } from "react-icons/fi"
import { PopupModal } from "../elements/PopupModal"
import { UserUpdateForm } from "../user/UserUpdateForm"
import { BsQrCode } from "react-icons/bs"
import { gql, useQuery } from "@apollo/client"
import { QueryLoading } from "./QueryLoading"
import ErrorMessage from "../ErrorMessage"
import moneyFormatter from "../../lib/moneyFormatter"
import { NoData } from "../elements/NoData"


enum DASH_STATE {
  DASHBOARD = 'dashboard',
  ORDERS = 'orders',
  SUBSCRIPTIONS = 'subscriptions',
  DOWNLOADS = 'downloads',
  TICKETS = 'tickets',
}


export function AccountDetails({ id, name, nameLast, email, tickets }: User) {

  const dashRef = useRef<HTMLHeadingElement | null>(null)
  const ordersRef = useRef<HTMLHeadingElement | null>(null)
  const subscriptionsRef = useRef<HTMLHeadingElement | null>(null)
  const downloadsRef = useRef<HTMLHeadingElement | null>(null)
  const ticketsRef = useRef<HTMLHeadingElement | null>(null)
  const [state, setState] = useState<string>(DASH_STATE.DASHBOARD)
  const [userData, setUserData] = useState<User>()
  const [bookingCells, setBookingCells] = useState([])
  const [subscriptionCells, setSubscriptionCells] = useState([])

  const { loading, error, data } = useQuery(
    USER_DASH_QUERY, {
    variables: { where: { id: id } }
  })

  function handleNavClick(state:string, ref:any){

    setState(state)

    if (ref.current) {
      // ref.current.scrollIntoView({ behavior: 'smooth' });
      // const scrollOffset = 50; // Adjust this value as needed
      // window.scrollBy(0, -scrollOffset);

      var headerOffset = 105;
      var elementPosition = ref.current.getBoundingClientRect().top;
      var offsetPosition = elementPosition + window.pageYOffset - headerOffset;
      
      window.scrollTo({
          top: offsetPosition,
          behavior: "smooth"
      }); 
    }
  };

  useEffect(() => {
    if(!data?.user) return

    const cellsBook = data.user.bookings.map((book:Booking) => ({
      date: datePrettyLocalDay(book.start || '') + ' ' + datePrettyLocalTime(book.start || ''),
      service: book.service.name,
      price: moneyFormatter(book.price),
      // end: datePrettyLocalDay(book.end || '') + ' ' + datePrettyLocalTime(book.end || ''),
      details: book.id,
    }))
    // console.log(cells);
    setBookingCells(cellsBook)

    const cellsSubs = data.user.subscriptions.map((sub:SubscriptionItem) => ({
      plan: sub.subscriptionPlan.name,
      status: sub.status,
      // end: datePrettyLocalDay(book.end || '') + ' ' + datePrettyLocalTime(book.end || ''),
      details: sub.id,
    }))
    // console.log(cells);
    setSubscriptionCells(cellsSubs)



    
    // return () => 
  }, [data?.user])

  if (loading) return <QueryLoading />
  if (error) return <ErrorMessage error={error} />
  // console.log(data);
  // const {user}:{user:User} = data

  return (<>
    <PopupModal data={userData} setData={setUserData}>
     <UserUpdateForm user={{ id, name, nameLast, email, tickets }} setUser={setUserData}/>
    </PopupModal>
    
    <StyledAccountCard>
      <StyledAccountNav>
        <ul>
          <li>
            <button 
              onClick={() => handleNavClick(DASH_STATE.DASHBOARD, dashRef)} 
              className={state === DASH_STATE.DASHBOARD ? 'active' : ''}
            >
              Dashboard <MdOutlineAccountBox />
            </button>
          </li>
          <li>
            <button 
              onClick={() => handleNavClick(DASH_STATE.ORDERS, ordersRef)} 
              className={state === DASH_STATE.ORDERS ? 'active' : ''}
            >
              Orders <MdShop />
            </button>
          </li>
          <li>
            <button 
              onClick={() => handleNavClick(DASH_STATE.SUBSCRIPTIONS, subscriptionsRef)} 
              className={state === DASH_STATE.SUBSCRIPTIONS ? 'active' : ''}
            >
              Subscriptions <MdAutorenew />
            </button>
          </li>
          <li>
            <button 
              onClick={() => handleNavClick(DASH_STATE.DOWNLOADS, downloadsRef)} 
              className={state === DASH_STATE.DOWNLOADS ? 'active' : ''}
            >
              Downloads <MdOutlineDownload />
            </button>
          </li>
          <li>
            <button 
              onClick={() => handleNavClick(DASH_STATE.TICKETS, ticketsRef)} 
              className={state === DASH_STATE.TICKETS ? 'active' : ''}
            >
              Tickets <HiOutlineTicket />
            </button>
          </li>
        </ul>
      </StyledAccountNav>

      <div className="dash-cont">

        <article ref={dashRef} className={state === DASH_STATE.DASHBOARD ? 'active' : ''}>
          <h3  >Dashboard</h3>
          <table>
            <tbody>
              <tr>
                <td>email: </td>
                <td>{email}</td>
              </tr>
              <tr>
                <td>name: </td>
                <td>{name} {nameLast}</td>
              </tr>
            </tbody>
          </table>
    
          <button data-tooltip={'edit'} onClick={() => setUserData({id: '123', name: 'yo', email: 'yo@bro.lan'})}>
            <FiEdit />
          </button>

        </article>

        <article ref={ordersRef} className={state === DASH_STATE.ORDERS ? 'active' : ''}>

          <h3>Orders / Services</h3>

          <Table 
            caption=""
            headers={[
              'service',
              'date',
              'price', 
              'details',
            ]}
            cells={bookingCells}
            route={`/bookings`}
          />

        </article>

        <article ref={subscriptionsRef} className={state === DASH_STATE.SUBSCRIPTIONS ? 'active' : ''}>
          <h3>Subscriptions</h3>
          <Table 
            caption=""
            headers={[
              'plan',
              'status',
              'details',
            ]}
            cells={subscriptionCells}
            route={`/subscriptions`}
          />

        </article>

        <article ref={downloadsRef} className={state === DASH_STATE.DOWNLOADS ? 'active' : ''}>
          <h3>Downloads</h3>

        </article>

        <article ref={ticketsRef} className={state === DASH_STATE.TICKETS ? 'active' : ''}>
          <h3> Tickets </h3>
          <StyledTicketList>
            
            {tickets && tickets.length <= 0 && <NoData /> }

            {tickets && tickets.map(tick => (
              <li key={tick.id}>
                <div className="meta">
                  <Link href={`/events/e/${tick.event?.id}`}>
                    <strong>{tick.event?.summary}</strong>
                  </Link>
                  <br />
                  <small>{datePrettyLocalDay(tick.event?.start || '')}</small>
                  <br />
                  <Link href={`/locations/${tick.event?.location?.id}`}>
                    <small>{tick.event?.location?.name}</small>
                  </Link>
                  <br />
                </div>

                <Link href={`/tickets/${tick.id}`} data-tooltip="ticket link QR code" className="button qrbutton"> 
                  <BsQrCode />
                </Link>

              </li>
            ))}
          </StyledTicketList>

        </article>
        

      </div>
    </StyledAccountCard>
  </> )
}

const USER_DASH_QUERY = gql`
  query User($where: UserWhereUniqueInput!) {
    user(where: $where) {
      bookings {
        id
        price
        start
        service {
          name
        }
        status
      }
      subscriptions{
        id
        subscriptionPlan {
          id
          name
        }
        status
      }
    }
  }
`

const StyledTicketList = styled.ul`
  padding: 0;

  li{
    border: solid 1px var(--c-primary);
    margin-bottom: .2em;
    border-radius: var(--br-dull);
    padding: .5em;

    display: flex;

    .qrbutton{
      margin-left: auto;
    }
  }
`

const StyledAccountCard = styled.div`
  display: flex;
  

  .dash-cont{
    flex-grow: 1;
    /* padding: 0 1em; */
    display: grid;

    h2{
      margin-top: 0;
    }

    article{
      /* height: 30em; */
      overflow-y: scroll;
      opacity: 0;
      transition: all .3s;
      background-color: var(--c-desaturated);
      padding: 0 1em;

      grid-column: 1;
      grid-row: 1;

      h3{
        transform: translateX(20px);
        transition: all .3s;
        text-decoration: underline;
      }
      
    }

    article.active{
      opacity: 1;
      /* pointer-events: all; */
      z-index: 1;

      h3{
        transform: translateX(0px);

      }
    }
  }

  @media screen and (max-width: 600px){
    nav{display: none}

    .dash-cont{
      article{
        opacity: 1;
        grid-column: auto;
        grid-row: auto;
        height: auto;
        overflow-y: visible;
        border-bottom: 1px solid black;
      }
    }
  }
`


const StyledAccountNav = styled.nav`
  /* background-color: #cbcbe2; */
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: start;

  > ul {
    position: sticky;
    top: 3rem;
  }
  

  li{
    display: flex;
    width: 100%;
  }
  
  
  button{
    flex-grow: 1;
    width: 100%;
    text-align: left;
    height: auto;
    align-self: stretch;
    padding: 2em 1em;
    text-transform: uppercase;
    font-size: .7rem;
    transition: all .3s;
    display: flex;
    align-items: center;
    justify-content: space-between;
    border-radius: 0;
    border: none;

    &.active{
      background-color: var(--c-primary);
      color: var(--c-txt);
    }

    svg {
      font-size: 1.6rem;
      float: right;
      margin-left: 1em;
    }
  }
`