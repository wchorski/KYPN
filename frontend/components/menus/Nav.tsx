import Link from 'next/link'
import styled from "styled-components";
import { SessionBadge, useUser } from "./Session";
import SignOutButton from './SignOutButton';
import { useCart } from '../../lib/cartState';
import CartCount from '../ecommerce/CartCount';
import { RiMenu3Line, RiMenu4Fill } from "react-icons/ri";
import { StyledNav, StyledUtilMenu } from '../../styles/menus/Nav.styled';
import { SearchInput } from './SearchInput';
import { MdClose, MdSearch } from 'react-icons/md';
import { useSearch } from '../../lib/useGlobalContext';
import { useNavControl } from '../../lib/useGlobalContext';
import { AnnouncementsMarquee } from '../elements/AnnouncementsMarquee';
import { EmojiFade } from '../elements/EmojiFade';
import { CartCount2 } from '../ecommerce/CartCount2';
import { LinkActive } from './LinkActive';

export function Nav() {

  const session = useUser()
  console.log({session});


  const { toggleCart, isOpen } = useCart()
  const { isSearchOpen, toggleSearch } = useSearch()
  const { isNavOpen, toggleNav } = useNavControl()

  return (
    <StyledNav className='main-menu' id='menu-main'>

      <SearchInput />


      <div id="menu-utility">

        <button
          onClick={e => toggleSearch()}
          className={isSearchOpen ? 'toggle-menu open' : 'toggle-menu'}
        >
          {isSearchOpen ? <MdClose /> : <MdSearch />}
        </button>

        {session && (<>
          <SessionBadge session={session} />
          <button
            onClick={e => toggleCart()}
            className={isOpen ? 'toggle-menu cart open' : 'toggle-menu cart'}
          >
            <CartCount2 count={session.cart.reduce(
              (tally: any, cartItem: any) => tally + (cartItem.product ? cartItem.quantity : 0),
              0
            )} />
          </button>
        </>)}

        <button
          onClick={e => toggleNav()}
          id='navwich'
          className={isNavOpen ? 'toggle-menu open' : 'toggle-menu'}
        >
          {isNavOpen ? <RiMenu4Fill /> : <RiMenu3Line />}
        </button>

      </div>

      <ul className={isNavOpen ? 'menu-main open' : 'menu-main '}>
        {session && session.isAdmin && (
          <LinkActive href='/admin'> Admin </LinkActive>
        )}
        <LinkActive href='/shop'> Shop </LinkActive>
        <LinkActive href='/blog'> Blog </LinkActive>
        <LinkActive href='/about'> About </LinkActive>
        <LinkActive href='/booking'> Book a Service </LinkActive>
        <LinkActive href='/events'> Live Events </LinkActive>

        {session && (<>
          <LinkActive href='/shop/sell'> Sell </LinkActive>
          <LinkActive href='/shop/orders'> Orders </LinkActive>
        </>)}

        {!session && (
          <LinkActive href='/auth/login'> Login </LinkActive>
        )}
      </ul>

    </StyledNav>
  )
}

