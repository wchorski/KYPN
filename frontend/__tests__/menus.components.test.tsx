import { render, screen, waitFor } from '@testing-library/react'
import { MockedProvider } from '@apollo/react-testing'
import { Nav } from '../components/Nav'
import { fakeUser, fakeCartItem } from '../lib/testUtils'
import { CartStateProvider } from '../lib/cartState'
import { QUERY_USER_CURRENT } from '../components/menus/Session'

const mocksNoLogin: any = [{
  request: {
    query: QUERY_USER_CURRENT,
    variables: {
      where: {
        id: '123',
      }
    }
  },
  result: {
    data: {
      authenticatedItem: null
    }
  }
}]

const mocksIsLoggedIn: any = [{
  request: {
    query: QUERY_USER_CURRENT,
    // variables: {
    //   where: {
    //     id: '123',
    //   }
    // }
  },
  result: {
    data: {
      authenticatedItem: fakeUser
    }
  }
}]

const mocksUserWithCart: any = [{
  request: {
    query: QUERY_USER_CURRENT,
    // variables: {
    //   where: {
    //     id: '4234',
    //   }
    // }
  },
  result: {
    data: {
      authenticatedItem: fakeUser({
        cart: [fakeCartItem, fakeCartItem, fakeCartItem]
      })
    }
  }
}]

describe('<Nav />', () => {

  it('renders minimal nav when no login', () => {

    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={mocksNoLogin} addTypename={false}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    )

    // debug()
    expect(container).toHaveTextContent('Login')
    expect(container).toMatchSnapshot()
    const link = screen.getByText('Login')
    // screen.debug(link)
    expect(link).toHaveAttribute('href', '/auth/login')

    const productsLink = screen.getByText('Shop')
    expect(productsLink).toBeInTheDocument()
    expect(productsLink).toHaveAttribute('href', '/shop')
  })


  it('renders nav when user islogged in', async () => {

    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={mocksIsLoggedIn} >
          <Nav />

        </MockedProvider>
      </CartStateProvider>
    )

    // TODO doesn't think it's logged back in
    // await screen.findByText('My Account')
    await waitFor(() => {
      // screen.findByText('My Account')
      // expect(container).toHaveTextContent('Sign Out')
      // debug()
    })
  })


  it('user with cart items', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={mocksUserWithCart} >
          <Nav />

        </MockedProvider>
      </CartStateProvider>
    )

    // await screen.findByText('My Account')
    debug()
  })

})