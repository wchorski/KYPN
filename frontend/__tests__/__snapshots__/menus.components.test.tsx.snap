// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> testing 18 items of pagination 1`] = `
<div>
  <div
    class="sc-a1cd700f-0 idA-DSX"
    data-testid="pagination"
  >
    <a
      aria-disabled="true"
      href="/shop/0"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
        />
      </svg>
      Prev
    </a>
    <span>
       
      1
       of 
      50
    </span>
    <span
      data-testid="pagination-countTotal"
    >
      1000
       Total Products
    </span>
    <a
      aria-disabled="false"
      href="/shop/2"
    >
      Next
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
        />
      </svg>
    </a>
  </div>
</div>
`;
